### ISSUE TITLE
go#lsp#DidChange[25] on save

### ISSUE THREAD

**Issue by mnarrell** [fatih/vim-go#3026]
### What did you do? (required: The issue will be **closed** when not provided)
I saved a file after making regular edits

### What happened instead?

|| Error detected while processing function
|| go#lsp#DidChange[25]
|| 33[47]
|| 36[11]
|| E900: Invalid channel id
|| E553: No more items
|| E553: No more items
|| E553: No more items
|| Error detected while processing function
|| go#auto#fmt_autosave[10]
|| go#fmt#Format[65]
|| go#fmt#update_file[20]
|| go#lsp#DidChange[25]
|| 33[47]
|| 36[11]
|| E900: Invalid channel id
|| "vss/resource_stream.go" 227L, 6309C
|| Error detected while processing function
|| go#auto#fmt_autosave[10]
|| go#fmt#Format[65]
|| go#fmt#update_file[20]
|| go#lsp#DidChange[25]
|| 33[47]
|| 36[11]
|| E900: Invalid channel id
|| "vss/resource_stream.go" 227L, 6309C
|| E748: No previously used register
|| Error detected while processing function
|| go#auto#fmt_autosave[10]
|| go#fmt#Format[65]
|| go#fmt#update_file[20]
|| go#lsp#DidChange[25]
|| 33[47]
|| 36[11]
|| E900: Invalid channel id
|| "vss/resource_stream.go" 227L, 6309C

This is now a consistent behavior until restarting vim.

####  `vimrc` you used to reproduce:

<details><summary>vimrc</summary><br><pre>
if has('python3')
    silent! python3 1
endif

source $XDG_CONFIG_HOME/nvim/autoload/plug.vim

call plug#begin('~/tmp/nvim/gogo')
Plug 'fatih/vim-go', { 'do': ':GoUpdateBinaries' }
call plug#end()
</pre></details>

#### Vim version (first three lines from `:version`):

NVIM v0.4.4
Build type: Release
LuaJIT 2.1.0-beta3

#### Go version (`go version`):

go version go1.15.1 darwin/amd64

#### Go environment
<details><summary><code>go env</code> Output:</summary><br><pre>

GO111MODULE="on"
GOARCH="amd64"
GOBIN="/Users/matt/src/work/go/bin"
GOCACHE="/Users/matt/Library/Caches/go-build"
GOENV="/Users/matt/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/matt/src/work/go/pkg/mod"
GOOS="darwin"
GOPATH="/Users/matt/src/work/go"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/matt/src/work/terraform-provider-vss/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/_1/_ttwj5gs5wz4z18pg26qqtx40000gn/T/go-build092074908=/tmp/go-build -gno-record-gcc-switches -fno-common"

</pre></details>

#### gopls version
<details><summary><code>gopls version</code> Output:</summary><br><pre>

golang.org/x/tools/gopls v0.5.0
    golang.org/x/tools/gopls@v0.5.0 h1:XEmO9RylgmaXp33iGrWfCGopVYDGBmLy+KmsIsfIo8Y=

</pre></details>

#### vim-go configuration:
<details><summary>vim-go configuration</summary><br><pre>
g:go_fmt_fail_silently = 1
g:go_loaded_gosnippets = 1
g:go_list_type = 'quickfix'
g:go_highlight_diagnostic_errors = 0
g:go_gopls_complete_unimported = 1
g:go_highlight_extra_types = 1
g:go_highlight_variable_declarations = 1
g:go_highlight_operators = 1
g:go_updatetime = 400
g:go_highlight_function_calls = 1
g:go_highlight_variable_assignments = 1
g:go_highlight_fields = 1
g:go_snippet_engine = 'ultisnips'
g:go_highlight_trailing_whitespace_error = 0
g:go_highlight_string_spellcheck = 1
g:go_echo_command_info = 0
g:go_highlight_function_parameters = 1
g:go_doc_popup_window = 1
g:go_highlight_format_strings = 1
g:go_highlight_structs = 1
g:go_highlight_chan_whitespace_error = 1
g:go_addtags_transform = 'camelcase'
g:go_highlight_functions = 1
g:go_gotags_bin = 'gotags'
g:go_highlight_space_tab_error = 0
g:go_highlight_types = 1
g:go_highlight_diagnostic_warnings = 0
g:go_fmt_command = 'gopls'
g:go_jump_to_error = 1
g:go_highlight_methods = 1
g:go_imports_autosave = 1
g:go_list_height = 20
g:go_implements_mode = 'gopls'
g:go_highlight_generate_tags = 1
g:go_highlight_build_constraints = 1
g:go_loaded_install = 1
g:go_highlight_array_whitespace_error = 1
</pre></details>

#### filetype detection configuration:
<details><summary>filetype detection</summary><br><pre>
filetype detection:ON  plugin:ON  indent:ON
</pre></details>

**Comment by bhcleek** [comment]
Can you provide the debug log for gopls by adding `let g:go_debug=["lsp"]` to your vimrc, duplicating the error, and pasting the debug output here?

Also, how can I duplicate this?

**Comment by mnarrell** [comment]
I'm not sure how to reproduce this reliably, but it happens often.  I'll try to capture some lsp output.

**Comment by mnarrell** [comment]
Ok, I've caught it, however the source content has sensitive info.  Is there a way to send this to you directly?

**Comment by mnarrell** [comment]
I believe this happens if i have one project with a Go module configuration I'm working on, within the same active VIM session, i open a file in a separate project with a different Go module configuration.  That kills the `gopls` server(s).

This just started with `gopls` `0.5.0`:

```
work/terraform-provider-vss master !
$ ps -ef | grep gopls
  501 75846 75811   0 12:54PM ??         0:00.23 /Users/matt/src/work/go/bin/gopls -debug localhost:0 -remote=auto
  501 75856 75846   0 12:54PM ??         0:04.90 /Users/matt/src/work/go/bin/gopls serve -listen unix;/var/folders/_1/_ttwj5gs5wz4z18pg26qqtx40000gn/T/gopls-1860b8-daemon.matt -listen.timeout 1m0s
  501 77505 54672   0  1:03PM ttys017    0:00.00 grep gopls

<!-- Here I navigate to second project -->

work/terraform-provider-vss master !
$ ps -ef | grep gopls
  501 77876 54672   0  1:04PM ttys017    0:00.00 grep gopls
```

**Comment by bhcleek** [comment]
That's not too surprising then. Try leveraging `:GoAddWorkspace` to add a new workspace all but the the first module you start editing.

@stamblerre is this related to https://github.com/golang/go/issues/32394 ?

**Comment by bhcleek** [comment]
> Ok, I've caught it, however the source content has sensitive info. Is there a way to send this to you directly?

You could send it to me as a file encrypted with my public SSH key or with gpg if you want.
* Here are my ssh public keys: https://api.github.com/users/bhcleek/keys.
* Here is my gpg public key: http://keys.gnupg.net/pks/lookup?search=bhcleek%40gmail.com&fingerprint=on&op=index

I also don't need the source code that may be in that debug log; you could elide it and provide the rest unencrypted here if you want.

FWIW, I think I can resolve at least part of this without waiting on a gopls fix by checking whether there's a workspace for the containing module when a file is opened and adding a workspace for the module if not.

**Comment by mnarrell** [comment]
I don't think we need to share the code now, this is in fact what's happening with switching workspaces.  Adding the second project via `GoAddWorkspace` addresses this issue.

**Comment by bhcleek** [comment]
Sounds good. I'm going to leave this open for now until I get it fixed so that you don't have to manually call `:GoAddWorkspace` for the use case you described.

**Comment by bhcleek** [comment]
@mnarrell can you confirm that you don't have to do `:GoAddWorkspace` explicitly with #3028  ?

**Comment by mnarrell** [comment]
I'm sorry, I don't know how to do this.  When I manually clone this repo, I only get the master branch.  If I try to add this hash to the vim-plug repo, I get:

```
x vim-go:
    fatal: reference is not a tree: 27843b8d19ce6588cab93f0ef4852502033a7582
```

**Comment by mnarrell** [comment]
Well, if I clone from your repo (bhcleek) instead of Fatih's, I can see this branch and yes, it looks like #3028 and gopls 0.5.0 resolve this issue.

**Comment by Shane-XB-Qian** [comment]
> can you confirm that you don't have to do `:GoAddWorkspace` explicitly with #3028 ?

sorry, jump in ..
but just tried sth else yesterday, i can confirm after your this change, some tools which depends on workspace had been fixed, so about `implicitly`, yes, it is, i do not need to manually do it now.

@bhcleek , but maybe silly i had 2 question,.. :)
1, was there a way to list what workspaces had been implicitly added ?
2, looks some tool still not work even implicitly added workspace, e.g 'GoKeyify', not sure if anyone really was using this ?

**Comment by bhcleek** [comment]
@Shane-XB-Qian I don't typically monitor closed issues.

Only `gopls` uses the workspaces; they have no relation to other tools. If you're seeing problems with `:GoKeyify` or other commands that do not leverage `gopls`, please open a separate issue.

**Comment by Shane-XB-Qian** [comment]
i was just about to edit previous comment, actually such 'implicitly' now is working for tools, but obvious not all cases, e.g omni compl still perhaps not working (and yes, it is 'gopls')
e.g : current outside path is '~/abc', the vim cwd is '~/abc/def/' (even this is the git repo root), and is editing '~/abc/def/ghi/xx.go', unless manually adding workspace with this ''~/abc/def/ghi/', omni not working.
but this is not a matter, just like you knew, and fyi. thanks your reply.

as for 'GoKeyify', ok, i will open a new ticket. thx.

**Comment by bhcleek** [comment]
@Shane-XB-Qian please open an issue describing how to duplicate the problems you're seeing with omni completion; I am unaware of any issues with gopls completion in the kind of scenario you seem to be describing.

**Comment by Shane-XB-Qian** [comment]
@bhcleek #3031 opened.
